### PR TITLE
Meta tensor channel

### PR DESCRIPTION
**Pull request by rijobro**
### Description
Channel-based transforms to use `MetaTensor`.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Hi @rijobro ,

Thanks for your quick update for MetaTensor support.
I feel maybe we don't need to remove the `NdarrayOrTensor` support for the current stage? It doesn't block the development of `MetaTensor` in transforms so far. And we already completed `NdarrayOrTensor` support for all transforms.
If really necessary to deprecate `NdarrayOrTensor`, we can remove it later?
@wyli @ericspod What do you think?

Thanks.

**Comment by rijobro** [comment]
It's true that `MetaTensor` is a subclass of `torch.Tensor`, which is a sub-part of `NdarrayOrTensor`, so we could leave it. But the proposed changes mean that the input to our transforms will become `torch.Tensor`, and so I don't think it's bad that these PRs start implementing that. It enables us to remove some `# type: ignore`, and should improve readability. 

I don't see the advantage of leaving them as `NdarrayOrTensor`, let me know your opinions on that one. Thanks

File: monai/data/image_reader.py [hunk 1]
```diff
@@ -833,8 +836,12 @@ def get_data(
         metadata["spatial_shape"] = np.asarray(region.shape[:-1])
         metadata["original_channel_dim"] = -1
 
+        # combine image and metadata
+        if not isinstance(region, MetaTensor):
+            region = MetaTensor(region, meta=deepcopy(metadata))
```
[the remark this answers is not in the document]
Comment (rijobro) [reply monai/data/image_reader.py]: Thanks. Deepcopy removed in https://github.com/Project-MONAI/MONAI/pull/4222/commits/4383f24c2cf1d36d3df4f17db0b4423e257ebb58.

**Comment by wyli** [comment]
I think I agree with @Nic-Ma, perhaps removing `NdarrayOrTensor` (and removing the numpy support) could be done in a separate dedicated PR? our previous discussion here https://github.com/Project-MONAI/MONAI/issues/3805 should be revisited...)

**Comment by rijobro** [comment]
Ok, how about this:
1. If a transform needs metadata (e.g., `EnsureChannelFirstd`), I will switch this to `torch.Tensor`, since it now can't be used with a `np.ndarray` as the metadata is no longer present,
2. If no metadata is required (e.g.,`AddChanneld`), I will leave this as `NdarrayOrTensor`.

Does that sounds reasonable?

**Comment by wyli** [comment]
> 

sounds good to me, what do you think @Nic-Ma ?

**Comment by Nic-Ma** [comment]
Hi @rijobro ,

Plan sounds good to me!
Thanks for the discussion.

File: monai/data/image_reader.py [hunk 2]
```diff
@@ -833,8 +834,13 @@ def get_data(
         metadata["spatial_shape"] = np.asarray(region.shape[:-1])
         metadata["original_channel_dim"] = -1
 
+        # combine image and metadata
+        if not isinstance(region, MetaTensor):
```
Comment (Nic-Ma) [inline monai/data/image_reader.py:838]: I think this wsi_reader is deprecated, maybe no need to update it?
We should update this part instead:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/wsi_reader.py#L234
@drbeh Could you please help double confirm it?

Thanks in advance.

File: monai/data/image_reader.py [hunk 3]
```diff
@@ -833,8 +834,13 @@ def get_data(
...
+            region = MetaTensor(region, meta=metadata)
+
         # Make it channel first
-        region = EnsureChannelFirst()(region, metadata)
+        region = EnsureChannelFirst()(region).numpy()
```
Comment (Nic-Ma) [inline monai/data/image_reader.py:842]: And may I know why cocnvert it to numpy array?

Thanks.

**Review by Nic-Ma — APPROVED** [review]
Overall looks good to me except for some minor comments inline.

Thanks.

File: monai/data/image_reader.py (hunk 3 above)
Comment (rijobro) [reply monai/data/image_reader.py:842]: Prior to these changes, it was a `np.ndarray` that gets passed through. Passing a `torch.Tensor` causes knock-on problems that I didn't think were worth spending the time to fix.

Comment (Nic-Ma) [reply monai/data/image_reader.py:842]: OK, I see your point, I think it's OK for the current stage, could you please help add a `FIXME` comment?

Thanks!

File: monai/data/image_reader.py (hunk 2 above)
Comment (drbeh) [reply monai/data/image_reader.py:838]: That's correct! `WSIReader` in `image_reader.py` is going to be deprecated in favor of the one in `wsi_reader.py`. Thanks

File: monai/data/image_reader.py (hunk 3 above)
Comment (rijobro) [reply monai/data/image_reader.py:843]: Done.

**Comment by Nic-Ma** [comment]
Hi @rijobro ,

As this PR is going to merge, could you please help add the doc-string for args of `update_meta`:
https://github.com/Project-MONAI/MONAI/blob/feature/MetaTensor/monai/data/meta_tensor.py#L117
I got confused before checking the code details.

Thanks in advance.

**Comment by rijobro** [comment]
> Hi @rijobro ,
As this PR is going to merge, could you please help add the doc-string for args of update_meta:
https://github.com/Project-MONAI/MONAI/blob/feature/MetaTensor/monai/data/meta_tensor.py#L117
I got confused before checking the code details.
Thanks in advance.

Sure @Nic-Ma, will do.

**Review by Nic-Ma — APPROVED** [review]

**Comment by rijobro** [comment]
@drbeh I needed to change [this line](https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/post/array.py#L736) from `self.filter.to(prob_map)` to self.filter.to(prob_map.device)`. Does this seem correct to you? Was the original code just to make sure the filter was on the same device as `prob_map`?

**Comment by drbeh** [comment]
> @drbeh I needed to change [this line](https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/post/array.py#L736) from `self.filter.to(prob_map)` to `self.filter.to(prob_map.device)`. Does this seem correct to you? Was the original code just to make sure the filter was on the same device as `prob_map`?

@rijobro you are right but `self.filter.to(prob_map)` is equivalent of `self.filter.to(prob_map.device).to(prob_map.dtype)` and I'm not sure if we need the `dtype` here. I can double check if you really want to change it.

**Review by wyli — APPROVED** [review]
mostly tested, merging it to trigger full integration tests.

**Pull request merged**